File: src/main/java/fu/se184491/loadmaster_be/entity/CenterOfGravity.java
package fu.se184491.loadmaster_be.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;

@Entity
@Table(name = "centers_of_gravity")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class CenterOfGravity {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "load_plan_id", unique = true, nullable = false)
    private LoadPlan loadPlan;

    @Column(name = "cog_x", precision = 8, scale = 2) private BigDecimal cogX;
    @Column(name = "cog_y", precision = 8, scale = 2) private BigDecimal cogY;
    @Column(name = "cog_z", precision = 8, scale = 2) private BigDecimal cogZ;
    @Column(name = "front_axle_load_kg", precision = 10, scale = 2) private BigDecimal frontAxleLoadKg;
    @Column(name = "rear_axle_load_kg", precision = 10, scale = 2) private BigDecimal rearAxleLoadKg;
    @Column(name = "is_balanced") private Boolean balanced;
}

File: OptimizeService/app/entity/optimization/center_of_gravity.py
from sqlalchemy import Column, Integer, BigInteger, ForeignKey, Numeric, Boolean
from sqlalchemy.orm import relationship

from app.config.database import Base


class CenterOfGravity(Base):
    __tablename__ = "centers_of_gravity"

    id = Column(Integer().with_variant(BigInteger, "postgresql"), primary_key=True, autoincrement=True)
    load_plan_id = Column(Integer().with_variant(BigInteger, "postgresql"), ForeignKey("load_plans.id"), unique=True, nullable=False)
    cog_x = Column(Numeric(8, 2), nullable=True)
    cog_y = Column(Numeric(8, 2), nullable=True)
    cog_z = Column(Numeric(8, 2), nullable=True)
    front_axle_load_kg = Column(Numeric(10, 2), nullable=True)
    rear_axle_load_kg = Column(Numeric(10, 2), nullable=True)
    is_balanced = Column(Boolean, default=True)

    load_plan = relationship("LoadPlan", backref="center_of_gravity")

    @property
    def is_axle_overload(self) -> bool:
        return not bool(self.is_balanced)

    @is_axle_overload.setter
    def is_axle_overload(self, value: bool) -> None:
        self.is_balanced = not value
